BSP/Simulator/Src/BSP_UART.c

```diff
@@ -50,104 +64,141 @@ void BSP_UART_Init(void) {
+    }
+    if(lineReceived2 && (usart == UART_BLE)) {
+        pthread_mutex_lock(&rx_mutex2); //NEED TO CHANGE THIS **********
+        RxFifo_Peek(&data, UART_BLE);
+        while(!RxFifo_IsEmpty(UART_BLE) && data != '\r' && data != '\n') {
+            recvd += RxFifo_Get((uint8_t *)str++, UART_BLE);
+            RxFifo_Peek(&data, UART_BLE);
+        }
+        RxFifo_Get(&data, UART_BLE);
+        *str = 0;
+        lineReceived2 = false;
+        pthread_mutex_unlock(&rx_mutex2);//NEED TO CHANG THIS *********
         return recvd;
     }
-
     return 0;
 }
 
 /**
  * @brief   Transmits data to through UART line
  * @param   str : pointer to buffer with data to send.
  * @param   len : size of buffer
+ * @param   usart : which usart to write to
  * @return  number of bytes that were sent
  */
-uint32_t BSP_UART_Write(char *str, uint32_t len) {
+uint32_t BSP_UART_Write(char *str, uint32_t len, UART_Port usart) {
     return printf("%.*s", len, str);
 }
 
 
-void *ScanThread(void *arg) {
-
+void *ScanThread3(void *arg) {
     do {
         int data = fgetc(stdin);
-
-        if(data == '\r' || data == '\n') {
-            lineReceived = true;
+        if(data == '\r' || data == '\n') lineReceived3 = true;
+        // Check if it was a backspace.
+        // '\b' for minicmom
+        // '\177' for putty
+        if(data != '\b' && data != '\177' && data != 0) {
+            pthread_mutex_lock(&rx_mutex3);
+            // Sweet, just a "regular" key. Put it into the fifo
+            // Doesn't matter if it fails. If it fails, then the data gets thrown away
+            // and the easiest solution for this is to increase RX_SIZE
+            RxFifo_Put((uint8_t)data, UART_USB);
+            pthread_mutex_unlock(&rx_mutex3);
         }
+    } while (1);
+    // Should not get here
+    int retval = 1;
+    pthread_exit(&retval);
+}
 
+void *ScanThread2(void *arg) {
+    do {
+        int data = fgetc(stdin);
+        if(data == '\r' || data == '\n') lineReceived2 = true;
         // Check if it was a backspace.
         // '\b' for minicmom
         // '\177' for putty
         if(data != '\b' && data != '\177' && data != 0) {
-
-            pthread_mutex_lock(&rx_mutex);
-
+            pthread_mutex_lock(&rx_mutex2);
             // Sweet, just a "regular" key. Put it into the fifo
             // Doesn't matter if it fails. If it fails, then the data gets thrown away
             // and the easiest solution for this is to increase RX_SIZE
-            RxFifo_Put((uint8_t)data);
-
-            pthread_mutex_unlock(&rx_mutex);
+            RxFifo_Put((uint8_t)data, UART_BLE);
+            pthread_mutex_unlock(&rx_mutex2);
         }
-
     } while (1);
-
     // Should not get here
     int retval = 1;
     pthread_exit(&retval);
 }
 
-static bool RxFifo_Get(uint8_t *data) {
-    if(!RxFifo_IsEmpty()) {
-        *data = rxBuffer[rxGet];
-        rxGet = (rxGet + 1) % RX_SIZE;
+static bool RxFifo_Get(uint8_t *data, UART_Port usart) {
```

Comment:
Where are you using the `UART_Port usart` parameter? I think this function will behave incorrectly if you try to read from the UART_BLE, but the UART_USB fifo also has data in it.